### PR TITLE
[clang] Ignore inline namespace for `hasName`

### PR DESCRIPTION
**Pull request by alejandro-alvarez-sonarsource**
Add a new flag `AlwaysSuppressInlineNamespace` to `PrintingPolicy` that is explicit
about *always* removing inline namespaces regardless of ambiguity.

Specializing a template from an inline namespace should be transparent. For instance

```
namespace foo {
    inline namespace v1 {
        template<typename A>
        void function(A&);
    }
}

namespace foo {
    template<>
    void function<int>(int&);
}
```

`hasName` should match both declarations of `foo::function`.

Makes the behavior of `matchesNodeFullSlow` and `matchesNodeFullFast` consistent, fixing
an assert inside `HasNameMatcher::matchesNode`.

### CODE REVIEW

**Review by erichkeane — COMMENTED** [review]
I have no problem with the code here, but don't know enough about ASTMatchers to be able to review this for 'well justified'. Do we have someone besides Aaron who does?

Also, it needs a release note I think for the matcher change.

**Comment by Sirraide** [comment]
I can’t help w/ AST matchers either unfortunately.

**Comment by alejandro-alvarez-sonarsource** [comment]
> Also, it needs a release note I think for the matcher change.

Added.

> I have no problem with the code here, but don't know enough about ASTMatchers to be able to review this for 'well justified'. Do we have someone besides Aaron who does?

FWIW this is an assertion error we saw downstream on debug builds with a snippet like the following (reduced), when trying to find `std::move`.

```cpp
namespace std {

inline namespace __1 {
template <class T> T &&move(T &&t) noexcept;

} // namespace __1
} // namespace std

namespace WTF {
enum CheckMoveParameterTag {};
}

namespace std {
template <WTF::CheckMoveParameterTag, typename T>
inline constexpr T &&move(T &&value) {
  return move(value);
}
} // namespace std

template <typename T> void f1(T t) { std::move(t); }

template void f1<int>(int);
```

The original code comes from [WebKit](https://github.com/qtwebkit/qtwebkit/blob/a60f9b3b9fc5236496f278fdf9df4855f41aab0f/PerformanceTests/StitchMarker/wtf/StdLibExtras.h#L502)

**Comment by alejandro-alvarez-sonarsource** [comment]
Ping (@PiotrZSL ?)

**Review by PiotrZSL — COMMENTED** [review]
Code looks fine.
Only thing is that existence of SuppressInlineNamespace and AlwaysSuppressInlineNamespace looks redundant.

Probably one should be named SuppressRedundantInlineNamespace and other SuppressAllInlineNamespace. Otherwise it looks strange. Or even SuppressInlineNamespace should be changed from bool to some enum like None, Redundant, All.

Make sure that all current test (specially from clang-tidy) still pass.

**Comment by AaronBallman** [comment]
> Or even SuppressInlineNamespace should be changed from bool to some enum like None, Redundant, All.

I think this is the suggestion I'd prefer; it's cleaner than adding a second boolean option.

**Comment by alejandro-alvarez-sonarsource** [comment]
> > Or even SuppressInlineNamespace should be changed from bool to some enum like None, Redundant, All.
> 
> I think this is the suggestion I'd prefer; it's cleaner than adding a second boolean option.

Agreed. And changes. 

> Make sure that all current test (specially from clang-tidy) still pass.

I have run both `check-clang` and `check-clang-tools` (locally) without failures.

**Comment by alejandro-alvarez-sonarsource** [comment]
I don't think the failure is related to this PR?

```
27: clang: error: unable to execute command: Executable "llvm-mc" doesn't exist!
```

File: clang/include/clang/AST/PrettyPrinter.h [hunk 1]
```diff
@@ -141,10 +143,11 @@ struct PrintingPolicy {
...
 
   /// Suppress printing parts of scope specifiers that correspond
-  /// to inline namespaces, where the name is unambiguous with the specifier
+  /// to inline namespaces.
+  /// If Redudant, where the name is unambiguous with the specifier removed.
+  /// If All, even if the name is ambiguous with the specifier
   /// removed.
-  LLVM_PREFERRED_TYPE(bool)
-  unsigned SuppressInlineNamespace : 1;
+  SuppressInlineNamespaceMode SuppressInlineNamespace : 2;
```
Comment (AaronBallman) [inline clang/include/clang/AST/PrettyPrinter.h:150]:
```suggestion
  LLVM_PREFERRD_TYPE(SuppressInlineNamespaceMode)
  unsigned SuppressInlineNamespace : 2;
```

File: clang/lib/AST/Decl.cpp [hunk 2]
```diff
@@ -1737,9 +1737,17 @@ void NamedDecl::printNestedNameSpecifier(raw_ostream &OS,
       continue;
 
     // Suppress inline namespace if it doesn't make the result ambiguous.
-    if (P.SuppressInlineNamespace && Ctx->isInlineNamespace() && NameInScope &&
-        cast<NamespaceDecl>(Ctx)->isRedundantInlineQualifierFor(NameInScope))
-      continue;
+    if (Ctx->isInlineNamespace() && NameInScope) {
+      bool const isRedundant =
```
Comment (AaronBallman) [inline clang/lib/AST/Decl.cpp:1741]:
```suggestion
      bool isRedundant =
```

File: clang/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -636,6 +636,9 @@ AST Matchers
 
 - Fixed a crash when traverse lambda expr with invalid captures. (#GH106444)
 
+- Ensure hasName matches template specializations across inline namespaces,
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:639]:
```suggestion
- Ensure ``hasName`` matches template specializations across inline namespaces,
```

**Review by AaronBallman — APPROVED** [review]
LGTM with some minor corrections, thank you!

And yeah, the precommit CI failure is unrelated to your changes.

Do you need someone to commit these changes for you? (If so, I can apply the changes myself and land them.)

File: clang/include/clang/AST/PrettyPrinter.h (hunk 1 above)
Comment (alejandro-alvarez-sonarsource) [reply clang/include/clang/AST/PrettyPrinter.h]: Applied by hand, since there was a typo (`PREFERRD` vs `PREFERRED`)

**Comment by alejandro-alvarez-sonarsource** [comment]
> Do you need someone to commit these changes for you? (If so, I can apply the changes myself and land them.)

Yes, please, I can't merge myself.

File: clang/include/clang/AST/PrettyPrinter.h (hunk 1 above)
Comment (AaronBallman) [reply clang/include/clang/AST/PrettyPrinter.h]: Oops, good catch! :-D

**Pull request merged**